Model/Observer/RecommendSettings.php

```diff
@@ -0,0 +1,216 @@
+    protected function validateRelatedProducts(string $changedPath): void
+    {
+        try {
+            $recommendations = $this->recommendManagement->getRelatedProductsRecommendation($this->getProductId());
+            if (empty($recommendations['renderingContent'])) {
+                $this->configWriter->save($changedPath, 0);
+                throw new LocalizedException(__(
+                    "It appears that there is no trained model available for Related Products recommendation for the AppID: %1.",
+                    $this->configHelper->getApplicationID()
+                ));
+            }
+        } catch (\Exception $e) {
+            $this->configWriter->save($changedPath, 0);
+            throw new LocalizedException(__($e->getMessage()));
+        }
+    }
+
+    /**
+     * @param string $changedPath
+     * @return void
+     * @throws LocalizedException
+     */
+    protected function validateTrendingItems(string $changedPath): void
+    {
+        try {
+            $recommendations = $this->recommendManagement->getTrendingItemsRecommendation();
+            // When no recommendations suggested, most likely trained model is missing
+            if (empty($recommendations['renderingContent'])) {
+                $this->configWriter->save($changedPath, 0);
+                throw new LocalizedException(__(
+                    "It appears that there is no trained model available for Trending Items recommendation for the AppID: %1.",
+                    $this->configHelper->getApplicationID()
+                ));
+            }
+        } catch (\Exception $e) {
+            $this->configWriter->save($changedPath, 0);
+            throw new LocalizedException(__($e->getMessage()));
+        }
+    }
+
+    /**
+     * @param string $changedPath
+     * @return void
+     * @throws LocalizedException
+     */
+    protected function validateLookingSimilar(string $changedPath): void
+    {
+        try {
+            $recommendations = $this->recommendManagement->getLookingSimilarRecommendation($this->getProductId());
+            if (empty($recommendations['renderingContent'])) {
+                $this->configWriter->save($changedPath, 0);
+                throw new LocalizedException(__(
+                    "It appears that there is no trained model available for Looking Similar Recommendation for the AppID: %1.",
+                    $this->configHelper->getApplicationID()
+                ));
+            }
+        } catch (\Exception $e) {
+            $this->configWriter->save($changedPath, 0);
+            throw new LocalizedException(__($e->getMessage()));
+        }
+    }
+
+    /**
+     * @return string
+     */
+    private function getProductId(): string
+    {
+        if ($this->productId === '') {
+            $searchCriteria = $this->searchCriteriaBuilder
+                ->addFilter('status', 1)
+                ->create();
+            $result = $this->productRepository->getList($searchCriteria);
```

Comment:
I'm concerned about this approach. This seems like a very non performant query especially for large catalogs. We should be applying `setPageSize()` and `setCurrentPage` to the `SearchCriteria`. We also are not accounting for out of stock or non visible products (and our new index non visible product setting). These could also be filters. I honestly think this could be abstracted as a service like `SampleProductFetcher`.

Reply:
I agree with out of stock and non visible filter should be added in the SearchCriteria to be in the safe side.  Adding setPageSize is also a good idea. It will put less stress on array_reverse and array_pop. I'll make that change in a few.